File: AOJ/Course/Algorithms and Data Structures I/Lesson 05 Recursion & Divide and Conquer/Merge Sort.rb
#!/usr/bin/env ruby
# TLE version

$SENTINEL = 0x3f3f3f3f3f
$tot = 0
n = gets.to_i
$x = gets.split.collect!{ |i| i.to_i }

def merge(left, mid, right)
	n1 = mid - left
	n2 = right - mid
	ll = $x[left, mid - left]
	rr = $x[mid, right - mid]
	ll << $SENTINEL
	rr << $SENTINEL
	i = 0
	j = 0
	$tot += right - left
	left.upto(right-1) do |k|
		if ll[i] <= rr[j]
			$x[k] = ll[i]
			i = i + 1
		else
			$x[k] = rr[j]
			j = j + 1
		end
	end
end

def msort(left, right)
	if left + 1 < right
		mid = (left + right) / 2
		msort(left, mid)
		msort(mid, right)
		merge(left, mid, right)
	end
end

msort(0, n)
puts $x.join(" ")
puts $tot

File: aoj/Course/Algorithms and Data Structures I/Lesson 05 Recursion & Divide and Conquer/Merge Sort.cc
#include <cstdio>
#include <cstring>
#include <algorithm>
using namespace std;

const int INF = 1000000000 + 10086;
const int MAXN = 500086;

int x[MAXN];
int tot;

void merge(int left, int mid, int right) {
	int n1 = mid - left, n2 = right - mid;
	int L[MAXN], R[MAXN];
	for (int i=0; i<n1; ++i) L[i] = x[left + i];
	for (int i=0; i<n2; ++i) R[i] = x[mid + i];
	L[n1] = INF, R[n2] = INF;
	int i = 0, j = 0;
	tot += right - left;
	for (int k=left; k<right; ++k) {
		if (L[i] <= R[j]) x[k] = L[i], ++i;
		else x[k] = R[j], ++j;
	}
}

void msort(int left, int right) {
	if (left + 1 < right) {
		int mid = (left + right) >> 1;
		msort(left, mid);
		msort(mid, right);
		merge(left, mid, right);
	}	
}

int main() {
	int n;

	scanf("%d", &n);
	for (int i=0; i<n; ++i) scanf("%d", x+i);
	msort(0, n);
	for (int i=0; i<n; ++i) printf("%d%s", x[i], i==n-1?"\n":" ");
	printf("%d\n", tot);

	return 0;
}
